cmd/deploy.go

```diff
@@ -0,0 +1,105 @@
+package cmd
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/docker/libcompose/docker"
+	"github.com/docker/libcompose/docker/ctx"
+	"github.com/docker/libcompose/project"
+	"github.com/spf13/cobra"
+)
+
+// deployCmd represents the up command
+var deployCmd = &cobra.Command{
+	Use:   "deploy",
+	Short: "Trigger an image deployment for all shipments and containers defined in compose files",
+	Long:  "Note that the deploy command is a subset of the up command without updates for environment variables, replicas, barge info, etc.",
+	Run:   deploy,
+}
+
+var shipmentBuildToken string
+
+func init() {
+	deployCmd.PersistentFlags().StringVarP(&shipmentBuildToken, "token", "t", "", "the shipment build token to use")
```

Comment:
Doh.  You're right.  Build tokens are at the shipment/environment level.  So for multi-shipment, you could share a docker-compose.yml, but you'd have to have a separate harbor-compose.yml files.  So here are the possible scenarios...

```
1) 1 shipment / 1 container (common)
  - may or may not be using compose
  - hc deploy

2) 1 shipment / many containers
  - hc deploy

3) 1 shipment / many containers / many environments
  - hc deploy
  - hc deploy -e qa -t ${QA_BUILD_TOKEN}

4) many shipments
  - hc deploy -c shipment1.yml
  - hc deploy -c shipment2.yml -t ${SHIPMENT2_TOKEN}

5) many shipments / many environments
  - hc deploy -c shipment1.yml
  - hc deploy -c shipment2.yml -t ${SHIPMENT2_TOKEN}
  - hc deploy -c shipment1.yml -t ${QA_BUILD_TOKEN}
  - hc deploy -c shipment2.yml -t ${SHIPMENT2_QA_BUILD_TOKEN}
```

So I still think this adds value and simplifies a CD script (by reducing repetitive boilerplate and leveraging existing compose files).  However, I'm not sure it's that much better than just adding a `deploy.yml` that defines all your deployment targets for all your shipments, containers, and environments and then just running:

```
docker-compose -f deploy.yml up
```

where `deploy.yml`:
```
version: '2'
services:
  s1:
    image: quay.io/turner/harbor-deploy
    command: [ "deploy", "--catalog=true" ]
    environment:
      ENVIRONMENT: dev
      SHIPMENT: shipment1
      CONTAINER: c1
      IMAGE: quay.io/turner/c1:${VERSION}
      TOKEN: ${SHIPMENT1_BUILD_TOKEN}
  s2:
    image: quay.io/turner/harbor-deploy
    command: [ "deploy", "--catalog=true" ]
    environment:
      ENVIRONMENT: dev
      SHIPMENT: shipment2
      CONTAINER: c2
      IMAGE: quay.io/turner/c2:${VERSION}
      TOKEN: ${SHIPMENT2_BUILD_TOKEN}
```

But it's a shame that environment, shipment, container, and image are all values that are already in your compose files (which also make building, testing, and pushing multiple containers easier).

Need to think on it a bit more...